File: core/src/main/java/com/graphhopper/routing/ev/StringEncodedValue.java
package com.graphhopper.routing.ev;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.*;

/**
 * This class holds a List of up to {@link #maxValues} encountered Strings and stores
 * <i>index+1</i> to indicate a string is set or <i>0</i> if no value is assigned
 *
 * @author Peter Karich
 * @author Thomas Butz
 */
public final class StringEncodedValue extends IntEncodedValueImpl {
    private final int maxValues;
    private final List<String> values;
    private final Map<String, Integer> indexMap;

    public StringEncodedValue(String name, int expectedValueCount) {
        this(name, expectedValueCount, false);
    }

    public StringEncodedValue(String name, int expectedValueCount, boolean storeTwoDirections) {
        super(name, 32 - Integer.numberOfLeadingZeros(expectedValueCount), storeTwoDirections);

        this.maxValues = roundUp(expectedValueCount);
        this.values = new ArrayList<>(maxValues);
        this.indexMap = new HashMap<>(maxValues);
    }

    public StringEncodedValue(String name, int bits, List<String> values, boolean storeTwoDirections) {
        super(name, bits, storeTwoDirections);

        this.maxValues = (1 << bits) - 1;
        if (values.size() > maxValues)
            throw new IllegalArgumentException("Number of values is higher than the maximum value count: "
                    + values.size() + " > " + maxValues);

        this.values = new ArrayList<>(values);
        this.indexMap = new HashMap<>(values.size());
        int index = 1;
        for (String value : values) {
            indexMap.put(value, index++);
        }
    }

    @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
    StringEncodedValue(
            @JsonProperty("name") String name,
            @JsonProperty("bits") int bits,
            @JsonProperty("min_storable_value") int minStorableValue,
            @JsonProperty("max_storable_value") int maxStorableValue,
            @JsonProperty("max_value") int maxValue,
            @JsonProperty("negate_reverse_direction") boolean negateReverseDirection,
            @JsonProperty("store_two_directions") boolean storeTwoDirections,
            @JsonProperty("fwd_data_index") int fwdDataIndex,
            @JsonProperty("bwd_data_index") int bwdDataIndex,
            @JsonProperty("fwd_shift") int fwdShift,
            @JsonProperty("bwd_shift") int bwdShift,
            @JsonProperty("fwd_mask") int fwdMask,
            @JsonProperty("bwd_mask") int bwdMask,
            @JsonProperty("max_values") int maxValues,
            @JsonProperty("values") List<String> values,
            @JsonProperty("index_map") HashMap<String, Integer> indexMap) {
        // we need this constructor for Jackson
        super(name, bits, minStorableValue, maxStorableValue, maxValue, negateReverseDirection, storeTwoDirections, fwdDataIndex, bwdDataIndex, fwdShift, bwdShift, fwdMask, bwdMask);
        if (values.size() > maxValues)
            throw new IllegalArgumentException("Number of values is higher than the maximum value count: "
                    + values.size() + " > " + maxValues);
        this.maxValues = maxValues;
        this.values = values;
        this.indexMap = indexMap;
    }

    public void setString(boolean reverse, int edgeId, EdgeBytesAccess edgeAccess, String value) {
        if (value == null) {
            super.setInt(reverse, edgeId, edgeAccess, 0);
            return;
        }
        int index = indexMap.getOrDefault(value, 0);
        if (index == 0) {
            if (values.size() == maxValues)
                throw new IllegalStateException("Maximum number of values reached for " + getName() + ": " + maxValues);

            values.add(value);
            index = values.size();
            indexMap.put(value, index);
        }
        super.setInt(reverse, edgeId, edgeAccess, index);
    }

    public String getString(boolean reverse, int edgeId, EdgeBytesAccess edgeAccess) {
        int value = super.getInt(reverse, edgeId, edgeAccess);
        if (value == 0) {
            return null;
        }
        return values.get(value - 1);
    }

    /**
     * @param value the value to be rounded
     * @return the value rounded to the highest integer with the same number of leading zeros
     */
    private static int roundUp(int value) {
        return -1 >>> Integer.numberOfLeadingZeros(value);
    }

    /**
     * @param value the String to retrieve the index
     * @return the non-zero index of the String or <i>0</i> if it couldn't be found
     */
    public int indexOf(String value) {
        return indexMap.getOrDefault(value, 0);
    }

    /**
     * @return an unmodifiable List of the current values
     */
    public List<String> getValues() {
        return Collections.unmodifiableList(values);
    }

}

File: core/src/main/kotlin/com/graphhopper/routing/ev/StringEncodedValue.kt
package com.graphhopper.routing.ev

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import java.util.Collections

/**
 * This class holds a List of up to [maxValues] encountered Strings and stores
 * *index+1* to indicate a string is set or *0* if no value is assigned
 *
 * @author Peter Karich
 * @author Thomas Butz
 */
class StringEncodedValue : IntEncodedValueImpl {
    // field names are part of the storage format (see IntEncodedValueImpl) — do not rename!
    private val maxValues: Int
    private val values: MutableList<String>
    private val indexMap: MutableMap<String, Int>

    constructor(name: String, expectedValueCount: Int) : this(name, expectedValueCount, false)

    constructor(name: String, expectedValueCount: Int, storeTwoDirections: Boolean) :
            super(name, 32 - Integer.numberOfLeadingZeros(expectedValueCount), storeTwoDirections) {
        this.maxValues = roundUp(expectedValueCount)
        this.values = ArrayList(maxValues)
        this.indexMap = HashMap(maxValues)
    }

    constructor(name: String, bits: Int, values: List<String>, storeTwoDirections: Boolean) :
            super(name, bits, storeTwoDirections) {
        this.maxValues = (1 shl bits) - 1
        if (values.size > maxValues)
            throw IllegalArgumentException("Number of values is higher than the maximum value count: " +
                    "${values.size} > $maxValues")

        this.values = ArrayList(values)
        this.indexMap = HashMap(values.size)
        var index = 1
        for (value in values) {
            indexMap[value] = index++
        }
    }

    @JsonCreator(mode = JsonCreator.Mode.PROPERTIES)
    internal constructor(
            @JsonProperty("name") name: String,
            @JsonProperty("bits") bits: Int,
            @JsonProperty("min_storable_value") minStorableValue: Int,
            @JsonProperty("max_storable_value") maxStorableValue: Int,
            @JsonProperty("max_value") maxValue: Int,
            @JsonProperty("negate_reverse_direction") negateReverseDirection: Boolean,
            @JsonProperty("store_two_directions") storeTwoDirections: Boolean,
            @JsonProperty("fwd_data_index") fwdDataIndex: Int,
            @JsonProperty("bwd_data_index") bwdDataIndex: Int,
            @JsonProperty("fwd_shift") fwdShift: Int,
            @JsonProperty("bwd_shift") bwdShift: Int,
            @JsonProperty("fwd_mask") fwdMask: Int,
            @JsonProperty("bwd_mask") bwdMask: Int,
            @JsonProperty("max_values") maxValues: Int,
            @JsonProperty("values") values: MutableList<String>,
            @JsonProperty("index_map") indexMap: HashMap<String, Int>) :
            super(name, bits, minStorableValue, maxStorableValue, maxValue, negateReverseDirection, storeTwoDirections,
                    fwdDataIndex, bwdDataIndex, fwdShift, bwdShift, fwdMask, bwdMask) {
        // we need this constructor for Jackson
        if (values.size > maxValues)
            throw IllegalArgumentException("Number of values is higher than the maximum value count: " +
                    "${values.size} > $maxValues")
        this.maxValues = maxValues
        this.values = values
        this.indexMap = indexMap
    }

    fun setString(reverse: Boolean, edgeId: Int, edgeIntAccess: EdgeIntAccess, value: String?) {
        if (value == null) {
            super.setInt(reverse, edgeId, edgeIntAccess, 0)
            return
        }
        var index = indexMap.getOrDefault(value, 0)
        if (index == 0) {
            if (values.size == maxValues)
                throw IllegalStateException("Maximum number of values reached for $name: $maxValues")

            values.add(value)
            index = values.size
            indexMap[value] = index
        }
        super.setInt(reverse, edgeId, edgeIntAccess, index)
    }

    fun getString(reverse: Boolean, edgeId: Int, edgeIntAccess: EdgeIntAccess): String? {
        val value = super.getInt(reverse, edgeId, edgeIntAccess)
        if (value == 0) {
            return null
        }
        return values[value - 1]
    }

    /**
     * @param value the String to retrieve the index
     * @return the non-zero index of the String or *0* if it couldn't be found
     */
    fun indexOf(value: String?): Int {
        return if (value == null) 0 else indexMap.getOrDefault(value, 0)
    }

    /**
     * @return an unmodifiable List of the current values
     */
    fun getValues(): List<String> = Collections.unmodifiableList(values)

    companion object {
        /**
         * @param value the value to be rounded
         * @return the value rounded to the highest integer with the same number of leading zeros
         */
        private fun roundUp(value: Int): Int = -1 ushr Integer.numberOfLeadingZeros(value)
    }
}
